src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,288 @@
+      throw std::invalid_argument("Diffusion parameter must be > 0");
+    }
+    if (hypers->q <= 0) {
+      throw std::invalid_argument("Number of factors must be > 0");
+    }
+  }
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void MFAHierarchy::update_hypers(
+    const std::vector<bayesmix::AlgorithmState::ClusterState>& states) {
+  auto& rng = bayesmix::Rng::Instance().get();
+  if (prior->has_fixed_values()) {
+    return;
+  }
+
+  else {
+    throw std::invalid_argument("Unrecognized hierarchy prior");
+  }
+}
+
+void MFAHierarchy::update_summary_statistics(const Eigen::RowVectorXd& datum,
+                                             bool add) {
+  if (add) {
+    data_sum += datum;
+    data.push_back(datum);
+  } else {
+    data_sum -= datum;
+    auto pos = std::find(data.begin(), data.end(), datum);
+    data.erase(pos);
+  }
+}
+
+void MFAHierarchy::clear_summary_statistics() {
+  data_sum = Eigen::VectorXd::Zero(p);
+  data.clear();
+}
+
+void MFAHierarchy::set_state_from_proto(
+    const google::protobuf::Message& state_) {
+  auto& statecast = downcast_state(state_);
+  state.mu = bayesmix::to_eigen(statecast.mfa_state().mu());
+  state.psi = bayesmix::to_eigen(statecast.mfa_state().psi());
+  state.Eta = bayesmix::to_eigen(statecast.mfa_state().eta());
+  state.Lambda = bayesmix::to_eigen(statecast.mfa_state().lambda());
+  set_card(statecast.cardinality());
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::ClusterState>
+MFAHierarchy::get_state_proto() const {
+  bayesmix::MFAState state_;
+  bayesmix::to_proto(state.mu, state_.mutable_mu());
+  bayesmix::to_proto(state.psi, state_.mutable_psi());
+  bayesmix::to_proto(state.Eta, state_.mutable_eta());
+  bayesmix::to_proto(state.Lambda, state_.mutable_lambda());
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::ClusterState>();
+  out->mutable_mfa_state()->CopyFrom(state_);
+  return out;
+}
+
+void MFAHierarchy::set_hypers_from_proto(
+    const google::protobuf::Message& hypers_) {
+  auto& hyperscast = downcast_hypers(hypers_).mfa_state();
+  hypers->mutilde = bayesmix::to_eigen(hyperscast.mutilde());
+  hypers->alpha0 = hyperscast.alpha0();
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  assert(update_params != true);  // should never be true
```

Comment:
1. Why is this the case?
2. Assertions are ignored when not compiling in debug mode, so this might not be what you want
3. `update_params` is currently always set as the negation of the `update_hierarchy_params()` method of the used  `Algorithm` class. In short, `update_params` solely depends on the chosen algorithm. Does this hierarchy require the creation of a new algorithm? If not, you should just ignore the value of `update_params` in this method

Reply:
We do not need to create a new algorithm, so we can just remove the assertion?

Reply:
I think so